Publish minutes of 2025-11-10 - 2025-11-14 TPAC meetings (#923)
hober/web-extensions: _minutes/2025-11-13-wecg-tpac.md

 * Secure Storage – 09:00 to 09:30
 * Clickjacking protections – 9:30 to 10:30
 * User Gestures – 11:00 to 11:45
 * Bug Triage – 11:45 to 12:30
 * Session 5 – 13:45 to 14:15
 * Sidebar API – 14:15 to 15:00

## Attendees

 * Rob Wu (Mozilla)
 * Simeon Vincent (Mozilla)
 * Kiara Rose (Apple)
 * Carlos Jeurissen (Independent)
 * Pattrick Kettner (Google Chrome)
 * Oliver Dunk (Google Chrome)
 * Devlin Cronin (Google Chrome)
 * Jordan Spivak (Capital One)
 * Casey Garland (Capital One)
 * George Henderson (Capital One)
 * Ricky Mondello (Apple)
 * Timothy Hatcher (Apple)
 * Rene Levelle (1Password)
 * Christian Brand (1Password)
 * Mukul Purohit (Microsoft)

## Notes

### Secure Storage

9:00 - 9:30

 * [rene] At 1Password we deal with a lot of secrets. Only way to work with this data currently is to store data behind a long password in a locked state. Ideally we'd use some form of biometry to get UV-backed storage access to encrypt and store this data in a browser extension context securely. Interested in reviving the secure storage proposal
   * https://github.com/w3c/webextensions/blob/main/proposals/secure-storage.md
 * [simeon] Can you clarify UV and UV-less storage?

>  * [rene] UV, user verification: touch ID, PIN entry on Windows, Windows Hello, etc.

Reviewer: ```suggestion
 * [rene] UV, user verification: Touch ID, PIN entry on Windows, Windows Hello, etc.
```

 * [rene] Whenever the user interacts with the page, there's no way of them telling whether the interaction is with the page itself or the content of the extension.
 * [rene] When an extension injects itself, extensions can try a number of methods to stay on top of other content: &lt;dialog> showModal, z-index fights, etc. These approaches aren't well catered to the type of UX we're trying to achieve here.
 * [rene] By design, pages can still display things above our UI. They can also just make a transparent element that's interactable above our UI.
 * [rene] Overlay detection is hard and fragile. Game of cat and mouse. We can do intersection observability, but there are a lot of false positives and it's expensive to run, slows down the UI. Mutation observes to detect content changes, but shadow roots counter. Monkeypatching, and the page monkey patches the patches.
 * [rene] Can't detect if we're blocking a malicious use of this API or if we're blocking a legitimate use. Can end up breaking websites.

>  * [christian] One thing we didn't have time to include is that there's been b-sides presentations, online blogs, various other places where researchers claim to have solutions. There are reasons we can't adopt some of those solutions. The extension's execution is governed by the page's CSP. Not possible to use an alert in all cases. Can't anticipate what all pages will do in the wild.

Reviewer: ```suggestion
 * [christian] One thing we didn't have time to include is that there's been BSides presentations, online blogs, various other places where researchers claim to have solutions. There are reasons we can't adopt some of those solutions. The extension's execution is governed by the page's CSP. Not possible to use an alert in all cases. Can't anticipate what all pages will do in the wild.
```

 * [ricky] If the page has no idea I'm going to do something because it doesn't have any insight into what I'm doing to prepare or display my UI… I don't know.
 * [oliver] There are also cases that we haven't discussed. Not security guarantees, but just wanting to draw on top of other content. Use cases like accessibility. We already have the concept of a top player. Order is entirely based on when content was promoted to the top layer, ordering is handled by the browser.
 * [george] Expect that would also create issues with extensions competing over the top layer.
 * [oliver] How concerned are we with extensions competing here? Is just being on top of page content a useful improvement?
 * [ricky] Yes.
 * [george] Guaranteeing being on the top, the last element in the page after the `<body>`.
 * [rob] Does that require mixing the dom from the content script and the web page? Any solution that mixes DOM of the page and the content is going to be difficult. A CSS transformation on the document element could also mess with rendering.
 * [ricky] I'm imaging the top layer as the stratosphere, and the extension layer as the top of the stratosphere. Is that accurate?
 * [oliver] Yes.

>  * [rob] Where does window.prompt() live? The web page could use that to take input?

Reviewer: ```suggestion
 * [rob] Where does `window.prompt()` live? The web page could use that to take input?
```

 * [christina] No typed text. We're not interested in changing how the user unlocks their extension context. Even if we have a secure layer, we don't want users to get used to entering their credentials in unsecure contexts. We're only interested in painting and click events.
 * [rob] On desktop you have openPopup(), on Firefox for Android, that completely replaces the tab. Is that a problem?
 * [christian] 1Password specific, we only have an extension on iOS. In that case we have native code execution, so we have other ways of entering credentials other than typing a password.
 * [carlos] To comment on this, on iOS the popup is an overlay.

>  * [ricky] window.prompt() was a fun one to think about.

Reviewer: ```suggestion
 * [ricky] `window.prompt()` was a fun one to think about.
```

 * [oliver] We're saying anything in the top layer from an extension draws on top of everything else.
 * [devlin] But you'd still potentially be subject to timing attack if the page has an idea of when they'll interact. Say there's a password field on the field. If the extension is listening on mouseup, the page registered a listener on mousedown and shows an “accept privacy policy”
 * [ricky] To expand on that example, the page listens on mouse down, the extension listens on mouse up.
 * [devlin] When I think about a timing attack as brought up before, the page knows when you're going to show your UI and where it will be presented. Whether the user is accepting a privacy policy or agreeing to play a game or whatever, the page puts a permission confirmation there instead. If the page can display its own UI in the right place a fraction of a second before the extension does…
 * [oliver] The timing attack you're describing is the user sees one UI and the click is long enough that they have time to see one UI they move their cursor and performs another click?
 * [devlin] The concern is that the page shows another UI before one password is able to do it

>  * …

Reviewer: ```suggestion
```

 * [rob] The concern is that after the user interacts with the page, there's no reliable way to ensure that the user is interacting with the extension. For argument, say after 10 seconds you know for sure that subsequent interactions will be with the extension.
 * [devlin] Couldn't you get those same guarantees by knowing you're on top and having an event to know when you're no longer on top?
 * [patrick] Yes but that feels convoluted.
 * [devlin] Point is that if we can find a solution that isn't extension specific.

>  * [rob] Such a method exists already: [document.elementFromPosition](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint), but that has not solved the issue.

Reviewer: ```suggestion
 * [rob] Such a method exists already: [`document.elementFromPosition`](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint), but that has not solved the issue.
```

>  * [] And no events to detect changes.

Reviewer: ```suggestion
 * [_unknown_] And no events to detect changes.
```

 * [devlin] Leaning towards any API that requires it consumes it.
 * [rob] Do you know whether extensions rely on currying user gestures?
 * [simeon] Absolutely.
 * [rob] For what use cases?
 * [devlin] Anything where a user interaction in the page requires user activation in the background.
 * [simeon] Concrete use case is requesting permissions from a content script.
 * [oliver] Would be useful to create issues to track these use cases. One would be ‘should we curry user gestures?'
 * [rob] For Firefox I'd like to make currying opt in, if we were to implement it.
 * [devlin] I expect that will break existing extensions and web pages.
 * [rob] Firefox does not support currying presently. And extensions should only opt in if they want to consume the user gesture.
 * [oliver] We should have another issue for user gestures in extensions being time based. There's a third one that slipped my mind at the moment.
 * [patrick] Definition of user activation in the browser.
 * [oliver] I think right now there's no
 * [devlin] Time, currying, consuming.
 * [oliver] I can open all of those.
 * [carlos]

>    * https://github.com/w3c/webextensions/issues/898

Reviewer: ```suggestion
 * [carlos] https://github.com/w3c/webextensions/issues/898
```

 * [timothy] Supportive of option for executeScript and sendMessage to support currying.
 * [devlin] Would you be open to a separate API and permission to synthesize a gesture in a web page?
 * [timothy] Possibly. We don't currently prompt for individual permissions. Could be a use case we entertain. Propagation sounds like it would solve 90+% of use cases.
 * [rob] Why do we want gestures without gestures?
 * [devlin] Agentic use cases. May need to synthesize a user gesture. Right now that's done with the debugger API.
 * [rob] Could be discussed, but want that to be separate, not blocking this.
 * [devlin] Understood and agreed.
 * [sarven] Thinking of scenarios where there's a conflict between actions on a web page vs. extensions. Say click in a page vs click in extensions code. What strategies are there to decide priority, whether one overrides the other, and, related, whether all the browsers do things consistently.
 * [devlin] We fire events in the order they were registered. That's not unique to user gestures, that's just event handling.
 * [sarven] Seems like there's a distinction between these cases.

>  * [rob] We have separate implementations for web platform APIs (addEventListener) and extensions APIs (addListener), etc.

Reviewer: ```suggestion
 * [rob] We have separate implementations for web platform APIs (`addEventListener`) and extensions APIs (`addListener`), etc.
```

   * [devlin] In Chrome it is more difficult technically to return a live object.
   * [carlos] Does Firefox have objections to a live object?
   * [rob] I do not have objections.
   * [carlos] Where is this on the timeline for Chromium?
   * [devlin] This is not a hill I would die on. A live object makes it less likely we would get to this earlier.
   * [carlos] Would you accept a patch using live objects?
   * [devlin] Sure.
   * [oliver] I will leave a comment saying we are happy with the new proposal.
   * [devlin] This is proposal A
   * **Resolution**: Consensus on Proposal A for getLanguageDictionary.
 * [Issue 873](https://github.com/w3c/webextensions/pull/873): Add specification for externally_connectable
   * [kiara] This is the specification for externally_connectable. There were some comments which I have addressed. There was a comment about it being unclear the two ways in which an extension can connect to another extension.
   * [devlin] If you don't specify externally_connectable, any extension can connect to your extension. I wonder how many extensions rely on this behavior.

>    * [rob] I'm sure extensions rely on it in Firefox. Tree Style Tabs uses this. Firefox supports onConnectExternal but not externally_connectable.

Reviewer: ```suggestion
   * [rob] I'm sure extensions rely on it in Firefox. Tree Style Tabs uses this. Firefox supports `onConnectExternal` but not `externally_connectable`.
```

>    * [timothy] We don't support externally_connectable with websites in Safari.

Reviewer 2: We _do_ support `externally_connectable` with websites in Safari. We _don't_ support it with other extensions.

 * [oliver] I thought if we closed the contextual side panel, it went back to the global side panel.
 * [carlos] There's two behaviors an extension developer may want. Always close or fall back. But you can't choose between the two if the API closes the global as fallback. While if you throw an exception when no contextual sidebar is present, a developer can still close the global if requested in the exception handling.
 * [rob] If an extension wants to close a contextual side panel, they pass a `tabId`. The expectation for it is to close the sidepanel.
 * [oliver] That would work as you described, and if the contextual sidepanel is not opened, the global one is closed instead.
 * [rob] As an extension developer, I would expect `close()` to close the sidepanel independently of what exactly is shown.
 * [oliver] Do we think sidepanel.open should match the behavior of close? I would prefer to keep the behavior consistent.
 * [patrick] If this is a breaking change, we can reach out to the affected developers. This group of developers is generally pretty engaged.
 * [oliver] I'm curious about your perspective, Kiara.

>  * [kiara] To me, it makes the most sense to not close and throw an error if called with a tab id. I don't necessarily think we should break things, but if we were doing it from scratch I'd want them to be consistent.

Reviewer: ```suggestion
 * [kiara] To me, it makes the most sense to not close and throw an error if called with a tab ID. I don't necessarily think we should break things, but if we were doing it from scratch I'd want them to be consistent.
```
